### DISCUSSION TITLE
Using the Eclipse Properties View with a GLSP Diagram

### DISCUSSION THREAD

**Discussion by Ginxss** [eclipse-glsp/glsp#806] [Q&A] [answered]
Hello,

I have a project which is based on the java-emf-eclipse project template (https://github.com/eclipse-glsp/glsp-examples/tree/master/project-templates/java-emf-eclipse).

When I select an element (a task in the example project), I want the Properties View in Eclipse to be populated with properties (e.g. the name of the task).

I found this tutorial quite helpful to give me a rough understanding of how it works: https://www.eclipse.org/articles/Article-Properties-View/properties-view.html

However, I just can't put the pieces together to make it work with my GLSP diagram. I am assuming there is an easy way to achieve what I'm looking for, since properties are an integral part of many diagrams.

Any help would be greatly appreciated.

**Comment by martin-fleck-at** [comment, marked as answer]
Hi @Ginxss,

Currently there is no default integration with the Eclipse Properties View but I think the general workflow would be something like this:

1. Ensure that you can retrieve the selection from your GLSP diagram. You have several options to do this: Register a custom action handler for the SelectAction that then sends the selection to the server using a new ServerSelectAction, listen to the existing selection service and then send the new custom action or override the selection service to achieve the same result. Probably having a custom action handler is the least intrusive for this but maybe you already have a customization in your project.

2. On the server side, register a ServerSelectActionhandler that publishes the selection (or an object based on it) to the ISelectionService of Eclipse. The Properties View listens to selection changes and aims to convert the selected object to an IPropertySource either by having the object already implemented the interface or adapt it to an IPropertySource or IPropertySourceProvider. So if the selection you provide fulfills these requirements, you should see something in the Property View.

I hope this gets you started.

**Reply by Ginxss** [reply]
Thanks for your answer @martin-fleck-at 

Unfortunately, this confused me even more. Why would I need to catch the SelectAction in the editor project and send a new event to the server when I could just catch the SelectAction in the server? Or did I misunderstand your point? I also couldn't even find a `ServerSelectAction`.

You usually need to register a selection provider in the `createPartControl()` method. The GLSP Diagram however has no selection provider registered. I tried to retroactively set a custom selection provider in my `SelectAction` handler (and then set `IPropertySource` object as the selection):

```java
GLSPDiagramEditor editor = GLSPIdeEditorPlugin.getDefaultGLSPEditorRegistry().getGLSPEditorOrThrow(modelState.getClientId());

if (editor.getSite().getSelectionProvider() == null) {
    editor.getSite().setSelectionProvider(selectionProvider);
}
```

This kind of works, but only "half" the time. The properties view is updated only if I click outside of the editor and then back into it, but not if I stay in the editor:

https://user-images.githubusercontent.com/13721331/207032239-2bb41927-ab96-48e2-b0d9-2ecbc6bd23c7.mp4

Maybe this is because the selection provider is not registered in the createPartControl() method? From [here](https://www.eclipse.org/articles/Article-WorkbenchSelections/article.html):

![2](https://user-images.githubusercontent.com/13721331/207032621-ddcf05a0-57a4-49fe-b435-b9569ddccd78.png)

Is there any way of getting a selection provider in the createPartControl() method of the GLSP Diagram?
Do you have any other ideas for what else I could do?

Thanks in advance!

**Reply by martin-fleck-at** [reply]
@Ginxss Sorry for the confusion! Yes, you can register the server to handle the `SelectAction`. It seems I misunderstood and though you still need some additional selection handling from the client in which case the easiest way would be to provide a custom action (e.g., a `ServerSelectAction`) that does not exist yet. 

As for the `createPartControl`: I believe you should be able to simply provide your own custom editor implementation that extends `GLSPDiagramEditor`. The default editor registry simply does an `instanceof` check so everything else should still work as expected. If you use an extension point to register your editor, you can simply adapt it with your own class:

```
   <extension
         point="org.eclipse.ui.editors">
      <editor
            class="org.eclipse.glsp.ide.editor.ui.GLSPDiagramEditor" <-- USE YOUR OWN CLASS HERE
            contributorClass="org.eclipse.glsp.ide.editor.ui.GLSPDiagramEditorActionContributor"
            default="true"
            extensions="wf"
            id="org.eclipse.glsp.workflow.editor"
            name="Workflow Diagram Editor">
      </editor>
   </extension>
```

**Reply by Ginxss** [reply]
@martin-fleck-at Thanks for your help!
This was actually the problem, and it works perfectly now with the custom editor implementation.

**Reply by Ginxss** [reply]
@martin-fleck-at Sorry to bother you again, but there is actually one more problem that I encountered.

When I update the properties of my model elements using the properties view, the model itself gets modified correctly, but in the diagram editor, nothing changes. Basically, the GModel is not rebuilt.
In order to update my visual diagram, I need to close and re-open the file.

This happens both when I use modelElement.eSet(...) or modelState.execute(command).

Is there some update function I can call on my editor or modelState to notify it that it has changed and that it should rebuild the GModel?

**Reply by martin-fleck-at** [reply]
@Ginxss You should be able to use the [ModelSubmissionHandler](https://github.com/eclipse-glsp/glsp-server/blob/master/plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/features/core/model/ModelSubmissionHandler.java) for this. It has two methods to send an update to the client-side model where the `submitModel` method also re-generates the GModel. Typically the submission handler is used when the model is [first requested](https://github.com/eclipse-glsp/glsp-server/blob/6d6903d8afcd61a2d0ce342bca6758ff45da83b7/plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/features/core/model/RequestModelActionHandler.java#L69), during [undo/redo](https://github.com/eclipse-glsp/glsp-server/blob/39ee6c9aa37f71d989ca0efcfc49ae2f25690b3e/plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/features/undoredo/UndoRedoActionHandler.java) or when [operations are invoked](https://github.com/eclipse-glsp/glsp-server/blob/0b9fbd55620e1728c78b0a83b7a432e638575686/plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/operations/OperationActionHandler.java#L66). Maybe the changing of the properties is not an Operation? Operations typically execute the command on the model state and submit the new model without you having to take care of it. If this does not work for you, you can try to submit it yourself through the ModelSubmissionHandler.

**Reply by Ginxss** [reply]
@martin-fleck-at Yes, the changing of the properties is currently being done directly in the `setPropertyValue` method from the `IPropertySource` object, and therefore is not an operation.
I instead tried using the `DefaultActionDispatcher` to send a new operation to the server. This works, but it gives me the warning "Discouraged Access". What would be the preferred way of dispatching operations?

**Reply by martin-fleck-at** [reply]
How are you retrieving the `DefaultActionDispatcher`? The `DefaultActionDispatcher` is defined in an internal package so if you inject the value, you should probably be using the interface `ActionDispatcher`.

**Comment by arcanefoam** [comment]
For anyone looking for additional implementation details, I wrote a [blog](https://kinori.tech/blog/en/2026/01/27/using-the-eclipse-properties-view-with-a-glsp-diagram/) explaining how you can use an existing Emf edit/editor to make this work:
